jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/ParallelGraphWriter.java

```diff
@@ -0,0 +1,371 @@
+ */
+
+package io.github.jbellis.jvector.graph.disk;
+
+import io.github.jbellis.jvector.disk.BufferedRandomAccessWriter;
+import io.github.jbellis.jvector.disk.RandomAccessWriter;
+import io.github.jbellis.jvector.graph.ImmutableGraphIndex;
+import io.github.jbellis.jvector.graph.disk.feature.Feature;
+import io.github.jbellis.jvector.graph.disk.feature.FeatureId;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousFileChannel;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.*;
+import java.util.function.IntFunction;
+
+/**
+ * Orchestrates parallel writing of L0 node records to disk using asynchronous file I/O.
+ * <p>
+ * This class manages:
+ * - A thread pool for building node records in parallel
+ * - Per-thread ImmutableGraphIndex.View instances for thread-safe neighbor iteration
+ * - A buffer pool to avoid excessive allocation
+ * - Asynchronous file channel writes that maintain correct ordering
+ * <p>
+ * Usage:
+ * <pre>
+ * try (var parallelWriter = new ParallelGraphWriter(writer, graph, config, filePath)) {
+ *     parallelWriter.writeL0Records(ordinalMapper, inlineFeatures, featureStateSuppliers, baseOffset);
+ * }
+ * </pre>
+ */
+class ParallelGraphWriter implements AutoCloseable {
+    private final RandomAccessWriter writer;
+    private final ImmutableGraphIndex graph;
+    private final ExecutorService executor;
+    private final ThreadLocal<ImmutableGraphIndex.View> viewPerThread;
+    private final ThreadLocal<ByteBuffer> bufferPerThread;
+    private final CopyOnWriteArrayList<ImmutableGraphIndex.View> allViews = new CopyOnWriteArrayList<>();
+    private final int recordSize;
+    private final Path filePath;
+
+    /**
+     * Configuration for parallel writing.
+     */
+    static class Config {
+        final int workerThreads;
+        final boolean useDirectBuffers;
+
+        /**
+         * @param workerThreads number of worker threads for building records (0 = use available processors)
+         * @param useDirectBuffers whether to use direct ByteBuffers (can be faster for large records)
+         */
+        public Config(int workerThreads, boolean useDirectBuffers) {
+            this.workerThreads = workerThreads <= 0 ? Runtime.getRuntime().availableProcessors() : workerThreads;
+            this.useDirectBuffers = useDirectBuffers;
+        }
+
+        /**
+         * Returns a default configuration suitable for most use cases.
+         * Uses available CPU cores and heap buffers.
+         *
+         * @return default configuration
+         */
+        public static Config defaultConfig() {
+            return new Config(0, false);
+        }
+    }
+
+    /**
+     * Creates a parallel writer.
+     *
+     * @param writer the underlying writer (must be BufferedRandomAccessWriter)
```

Comment:
`the underlying writer (must be BufferedRandomAccessWriter)` does that mean we need to limit the type to `BufferedRandomAccessWriter`?